// @payloadcms/db-postgres/dist/connect.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "connect", {
    enumerable: true,
    get: function() {
        return connect;
    }
});
const _drizzleorm = require("drizzle-orm");
const _nodepostgres = require("drizzle-orm/node-postgres");
const _pg = require("pg");
const _pushDevSchema = require("./utilities/pushDevSchema");
const connectWithReconnect = async function({ adapter, payload, reconnect = false }) {
    let result;
    if (!reconnect) {
        result = await adapter.pool.connect();
    } else {
        try {
            result = await adapter.pool.connect();
        } catch (err) {
            setTimeout(()=>{
                payload.logger.info('Reconnecting to postgres');
                void connectWithReconnect({
                    adapter,
                    payload,
                    reconnect: true
                });
            }, 1000);
        }
    }
    if (!result) {
        return;
    }
    result.prependListener('error', (err)=>{
        try {
            if (err.code === 'ECONNRESET') {
                void connectWithReconnect({
                    adapter,
                    payload,
                    reconnect: true
                });
            }
        } catch (err) {
        // swallow error
        }
    });
};
const connect = async function connect(payload) {
    this.schema = {
        pgSchema: this.pgSchema,
        ...this.tables,
        ...this.relations,
        ...this.enums
    };
    try {
        this.pool = new _pg.Pool(this.poolOptions);
        await connectWithReconnect({
            adapter: this,
            payload
        });
        const logger = this.logger || false;
        this.drizzle = (0, _nodepostgres.drizzle)(this.pool, {
            logger,
            schema: this.schema
        });
        if (process.env.PAYLOAD_DROP_DATABASE === 'true') {
            this.payload.logger.info(`---- DROPPING TABLES SCHEMA(${this.schemaName || 'public'}) ----`);
            await this.drizzle.execute(_drizzleorm.sql.raw(`
        drop schema if exists ${this.schemaName || 'public'} cascade;
        create schema ${this.schemaName || 'public'};
      `));
            this.payload.logger.info('---- DROPPED TABLES ----');
        }
    } catch (err) {
        payload.logger.error(`Error: cannot connect to Postgres. Details: ${err.message}`, err);
        process.exit(1);
    }
    // Only push schema if not in production
    if (process.env.NODE_ENV !== 'production' && process.env.PAYLOAD_MIGRATING !== 'true' && this.push !== false) {
        await (0, _pushDevSchema.pushDevSchema)(this);
    }
};
