// @payloadcms/db-postgres/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "postgresAdapter", {
    enumerable: true,
    get: function() {
        return postgresAdapter;
    }
});
const _fs = /*#__PURE__*/ _interop_require_default(require("fs"));
const _path = /*#__PURE__*/ _interop_require_default(require("path"));
const _database = require("payload/database");
const _connect = require("./connect");
const _count = require("./count");
const _create = require("./create");
const _createGlobal = require("./createGlobal");
const _createGlobalVersion = require("./createGlobalVersion");
const _createMigration = require("./createMigration");
const _createVersion = require("./createVersion");
const _deleteMany = require("./deleteMany");
const _deleteOne = require("./deleteOne");
const _deleteVersions = require("./deleteVersions");
const _destroy = require("./destroy");
const _find = require("./find");
const _findGlobal = require("./findGlobal");
const _findGlobalVersions = require("./findGlobalVersions");
const _findOne = require("./findOne");
const _findVersions = require("./findVersions");
const _init = require("./init");
const _migrate = require("./migrate");
const _migrateDown = require("./migrateDown");
const _migrateFresh = require("./migrateFresh");
const _migrateRefresh = require("./migrateRefresh");
const _migrateReset = require("./migrateReset");
const _migrateStatus = require("./migrateStatus");
const _queryDrafts = require("./queryDrafts");
const _beginTransaction = require("./transactions/beginTransaction");
const _commitTransaction = require("./transactions/commitTransaction");
const _rollbackTransaction = require("./transactions/rollbackTransaction");
const _update = require("./update");
const _updateGlobal = require("./updateGlobal");
const _updateGlobalVersion = require("./updateGlobalVersion");
const _updateVersion = require("./updateVersion");
function _interop_require_default(obj) {
    return obj && obj.__esModule ? obj : {
        default: obj
    };
}
function postgresAdapter(args) {
    const postgresIDType = args.idType || 'serial';
    const payloadIDType = postgresIDType === 'serial' ? 'number' : 'text';
    function adapter({ payload }) {
        const migrationDir = findMigrationDir(args.migrationDir);
        return (0, _database.createDatabaseAdapter)({
            name: 'postgres',
            drizzle: undefined,
            enums: {},
            fieldConstraints: {},
            idType: postgresIDType,
            localesSuffix: args.localesSuffix || '_locales',
            logger: args.logger,
            pgSchema: undefined,
            pool: undefined,
            poolOptions: args.pool,
            push: args.push,
            relations: {},
            relationshipsSuffix: args.relationshipsSuffix || '_rels',
            schema: {},
            schemaName: args.schemaName,
            sessions: {},
            tableNameMap: new Map(),
            tables: {},
            versionsSuffix: args.versionsSuffix || '_v',
            // DatabaseAdapter
            beginTransaction: _beginTransaction.beginTransaction,
            commitTransaction: _commitTransaction.commitTransaction,
            connect: _connect.connect,
            count: _count.count,
            create: _create.create,
            createGlobal: _createGlobal.createGlobal,
            createGlobalVersion: _createGlobalVersion.createGlobalVersion,
            createMigration: _createMigration.createMigration,
            createVersion: _createVersion.createVersion,
            defaultIDType: payloadIDType,
            deleteMany: _deleteMany.deleteMany,
            deleteOne: _deleteOne.deleteOne,
            deleteVersions: _deleteVersions.deleteVersions,
            destroy: _destroy.destroy,
            find: _find.find,
            findGlobal: _findGlobal.findGlobal,
            findGlobalVersions: _findGlobalVersions.findGlobalVersions,
            findOne: _findOne.findOne,
            findVersions: _findVersions.findVersions,
            init: _init.init,
            migrate: _migrate.migrate,
            migrateDown: _migrateDown.migrateDown,
            migrateFresh: _migrateFresh.migrateFresh,
            migrateRefresh: _migrateRefresh.migrateRefresh,
            migrateReset: _migrateReset.migrateReset,
            migrateStatus: _migrateStatus.migrateStatus,
            migrationDir,
            payload,
            queryDrafts: _queryDrafts.queryDrafts,
            rollbackTransaction: _rollbackTransaction.rollbackTransaction,
            updateGlobal: _updateGlobal.updateGlobal,
            updateGlobalVersion: _updateGlobalVersion.updateGlobalVersion,
            updateOne: _update.updateOne,
            updateVersion: _updateVersion.updateVersion
        });
    }
    return adapter;
}
/**
 * Attempt to find migrations directory.
 *
 * Checks for the following directories in order:
 * - `migrationDir` argument from Payload config
 * - `src/migrations`
 * - `dist/migrations`
 * - `migrations`
 *
 * Defaults to `src/migrations`
 *
 * @param migrationDir
 * @returns
 */ function findMigrationDir(migrationDir) {
    const cwd = process.cwd();
    const srcDir = _path.default.resolve(cwd, 'src/migrations');
    const distDir = _path.default.resolve(cwd, 'dist/migrations');
    const relativeMigrations = _path.default.resolve(cwd, 'migrations');
    // Use arg if provided
    if (migrationDir) return migrationDir;
    // Check other common locations
    if (_fs.default.existsSync(srcDir)) {
        return srcDir;
    }
    if (_fs.default.existsSync(distDir)) {
        return distDir;
    }
    if (_fs.default.existsSync(relativeMigrations)) {
        return relativeMigrations;
    }
    return srcDir;
}
