// @payloadcms/richtext-lexical/dist/index.js
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
function _export(target, all) {
    for(var name in all)Object.defineProperty(target, name, {
        enumerable: true,
        get: all[name]
    });
}
_export(exports, {
    $createAutoLinkNode: function() {
        return _AutoLinkNode.$createAutoLinkNode;
    },
    $createBlockNode: function() {
        return _BlocksNode.$createBlockNode;
    },
    $createLinkNode: function() {
        return _LinkNode.$createLinkNode;
    },
    $createRelationshipNode: function() {
        return _RelationshipNode.$createRelationshipNode;
    },
    $createUploadNode: function() {
        return _UploadNode.$createUploadNode;
    },
    $isAutoLinkNode: function() {
        return _AutoLinkNode.$isAutoLinkNode;
    },
    $isBlockNode: function() {
        return _BlocksNode.$isBlockNode;
    },
    $isLinkNode: function() {
        return _LinkNode.$isLinkNode;
    },
    $isRelationshipNode: function() {
        return _RelationshipNode.$isRelationshipNode;
    },
    $isUploadNode: function() {
        return _UploadNode.$isUploadNode;
    },
    AlignFeature: function() {
        return _align.AlignFeature;
    },
    AutoLinkNode: function() {
        return _AutoLinkNode.AutoLinkNode;
    },
    BlockNode: function() {
        return _BlocksNode.BlockNode;
    },
    BlockQuoteFeature: function() {
        return _BlockQuote.BlockQuoteFeature;
    },
    BlocksFeature: function() {
        return _Blocks.BlocksFeature;
    },
    BoldTextFeature: function() {
        return _Bold.BoldTextFeature;
    },
    CAN_USE_DOM: function() {
        return _canUseDOM.CAN_USE_DOM;
    },
    CheckListFeature: function() {
        return _CheckList.CheckListFeature;
    },
    DETAIL_TYPE_TO_DETAIL: function() {
        return _nodeFormat.DETAIL_TYPE_TO_DETAIL;
    },
    DOUBLE_LINE_BREAK: function() {
        return _nodeFormat.DOUBLE_LINE_BREAK;
    },
    ELEMENT_FORMAT_TO_TYPE: function() {
        return _nodeFormat.ELEMENT_FORMAT_TO_TYPE;
    },
    ELEMENT_TYPE_TO_FORMAT: function() {
        return _nodeFormat.ELEMENT_TYPE_TO_FORMAT;
    },
    ENABLE_SLASH_MENU_COMMAND: function() {
        return _index1.ENABLE_SLASH_MENU_COMMAND;
    },
    EditorConfigProvider: function() {
        return _EditorConfigProvider.EditorConfigProvider;
    },
    FormatSectionWithEntries: function() {
        return _floatingSelectToolbarSection.SectionWithEntries;
    },
    HTMLConverterFeature: function() {
        return _html.HTMLConverterFeature;
    },
    HeadingFeature: function() {
        return _Heading.HeadingFeature;
    },
    HorizontalRuleFeature: function() {
        return _horizontalrule.HorizontalRuleFeature;
    },
    IS_ALL_FORMATTING: function() {
        return _nodeFormat.IS_ALL_FORMATTING;
    },
    IndentFeature: function() {
        return _indent.IndentFeature;
    },
    InlineCodeTextFeature: function() {
        return _InlineCode.InlineCodeTextFeature;
    },
    ItalicTextFeature: function() {
        return _Italic.ItalicTextFeature;
    },
    LTR_REGEX: function() {
        return _nodeFormat.LTR_REGEX;
    },
    LexicalPluginToLexicalFeature: function() {
        return _LexicalPluginToLexical.LexicalPluginToLexicalFeature;
    },
    LinebreakHTMLConverter: function() {
        return _linebreak.LinebreakHTMLConverter;
    },
    LinkFeature: function() {
        return _Link.LinkFeature;
    },
    LinkNode: function() {
        return _LinkNode.LinkNode;
    },
    NON_BREAKING_SPACE: function() {
        return _nodeFormat.NON_BREAKING_SPACE;
    },
    NodeFormat: function() {
        return _nodeFormat.NodeFormat;
    },
    OrderedListFeature: function() {
        return _OrderedList.OrderedListFeature;
    },
    ParagraphFeature: function() {
        return _Paragraph.ParagraphFeature;
    },
    ParagraphHTMLConverter: function() {
        return _paragraph.ParagraphHTMLConverter;
    },
    Point: function() {
        return _point.Point;
    },
    RTL_REGEX: function() {
        return _nodeFormat.RTL_REGEX;
    },
    RawUploadPayload: function() {
        return _UploadNode.RawUploadPayload;
    },
    Rect: function() {
        return _rect.Rect;
    },
    RelationshipFeature: function() {
        return _Relationship.RelationshipFeature;
    },
    RelationshipNode: function() {
        return _RelationshipNode.RelationshipNode;
    },
    SlashMenuGroup: function() {
        return _types.SlashMenuGroup;
    },
    SlashMenuOption: function() {
        return _types.SlashMenuOption;
    },
    SlateBlockquoteConverter: function() {
        return _blockquote.SlateBlockquoteConverter;
    },
    SlateHeadingConverter: function() {
        return _heading.SlateHeadingConverter;
    },
    SlateIndentConverter: function() {
        return _indent1.SlateIndentConverter;
    },
    SlateLinkConverter: function() {
        return _link.SlateLinkConverter;
    },
    SlateListItemConverter: function() {
        return _listItem.SlateListItemConverter;
    },
    SlateOrderedListConverter: function() {
        return _orderedList.SlateOrderedListConverter;
    },
    SlateRelationshipConverter: function() {
        return _relationship.SlateRelationshipConverter;
    },
    SlateToLexicalFeature: function() {
        return _SlateToLexical.SlateToLexicalFeature;
    },
    SlateUnknownConverter: function() {
        return _unknown.SlateUnknownConverter;
    },
    SlateUnorderedListConverter: function() {
        return _unorderedList.SlateUnorderedListConverter;
    },
    SlateUploadConverter: function() {
        return _upload.SlateUploadConverter;
    },
    StrikethroughTextFeature: function() {
        return _strikethrough.StrikethroughTextFeature;
    },
    SubscriptTextFeature: function() {
        return _subscript.SubscriptTextFeature;
    },
    SuperscriptTextFeature: function() {
        return _superscript.SuperscriptTextFeature;
    },
    TEXT_MODE_TO_TYPE: function() {
        return _nodeFormat.TEXT_MODE_TO_TYPE;
    },
    TEXT_TYPE_TO_FORMAT: function() {
        return _nodeFormat.TEXT_TYPE_TO_FORMAT;
    },
    TEXT_TYPE_TO_MODE: function() {
        return _nodeFormat.TEXT_TYPE_TO_MODE;
    },
    TOGGLE_LINK_COMMAND: function() {
        return _LinkNode.TOGGLE_LINK_COMMAND;
    },
    TestRecorderFeature: function() {
        return _TestRecorder.TestRecorderFeature;
    },
    TextDropdownSectionWithEntries: function() {
        return _floatingSelectToolbarTextDropdownSection.TextDropdownSectionWithEntries;
    },
    TextHTMLConverter: function() {
        return _text.TextHTMLConverter;
    },
    TreeViewFeature: function() {
        return _TreeView.TreeViewFeature;
    },
    UnderlineTextFeature: function() {
        return _underline.UnderlineTextFeature;
    },
    UnorderedListFeature: function() {
        return _UnorderedList.UnorderedListFeature;
    },
    UploadFeature: function() {
        return _Upload.UploadFeature;
    },
    UploadNode: function() {
        return _UploadNode.UploadNode;
    },
    addSwipeDownListener: function() {
        return _swipe.addSwipeDownListener;
    },
    addSwipeLeftListener: function() {
        return _swipe.addSwipeLeftListener;
    },
    addSwipeRightListener: function() {
        return _swipe.addSwipeRightListener;
    },
    addSwipeUpListener: function() {
        return _swipe.addSwipeUpListener;
    },
    cloneDeep: function() {
        return _cloneDeep.cloneDeep;
    },
    consolidateHTMLConverters: function() {
        return _field.consolidateHTMLConverters;
    },
    convertLexicalNodesToHTML: function() {
        return _converter.convertLexicalNodesToHTML;
    },
    convertLexicalToHTML: function() {
        return _converter.convertLexicalToHTML;
    },
    convertSlateNodesToLexical: function() {
        return _index.convertSlateNodesToLexical;
    },
    convertSlateToLexical: function() {
        return _index.convertSlateToLexical;
    },
    createBlockNode: function() {
        return _createBlockNode.createBlockNode;
    },
    defaultEditorConfig: function() {
        return _default.defaultEditorConfig;
    },
    defaultEditorFeatures: function() {
        return _default.defaultEditorFeatures;
    },
    defaultHTMLConverters: function() {
        return _defaultConverters.defaultHTMLConverters;
    },
    defaultRichTextValue: function() {
        return _defaultValue.defaultRichTextValue;
    },
    defaultSanitizedEditorConfig: function() {
        return _default.defaultSanitizedEditorConfig;
    },
    defaultSlateConverters: function() {
        return _defaultConverters1.defaultSlateConverters;
    },
    getDOMRangeRect: function() {
        return _getDOMRangeRect.getDOMRangeRect;
    },
    getEnabledNodes: function() {
        return _nodes.getEnabledNodes;
    },
    getSelectedNode: function() {
        return _getSelectedNode.getSelectedNode;
    },
    invariant: function() {
        return _invariant.invariant;
    },
    isHTMLElement: function() {
        return _guard.isHTMLElement;
    },
    isPoint: function() {
        return _point.isPoint;
    },
    joinClasses: function() {
        return _joinClasses.joinClasses;
    },
    lexicalEditor: function() {
        return lexicalEditor;
    },
    lexicalHTML: function() {
        return _field.lexicalHTML;
    },
    loadFeatures: function() {
        return _loader.loadFeatures;
    },
    sanitizeEditorConfig: function() {
        return _sanitize.sanitizeEditorConfig;
    },
    sanitizeFeatures: function() {
        return _sanitize.sanitizeFeatures;
    },
    sanitizeUrl: function() {
        return _url.sanitizeUrl;
    },
    setFloatingElemPosition: function() {
        return _setFloatingElemPosition.setFloatingElemPosition;
    },
    setFloatingElemPositionForLinkEditor: function() {
        return _setFloatingElemPositionForLinkEditor.setFloatingElemPositionForLinkEditor;
    },
    sortFeaturesForOptimalLoading: function() {
        return _loader.sortFeaturesForOptimalLoading;
    },
    useEditorConfigContext: function() {
        return _EditorConfigProvider.useEditorConfigContext;
    },
    validateUrl: function() {
        return _url.validateUrl;
    }
});
const _utilities = require("payload/utilities");
const _default = require("./field/lexical/config/default");
const _sanitize = require("./field/lexical/config/sanitize");
const _cloneDeep = require("./field/lexical/utils/cloneDeep");
const _richTextRelationshipPromise = require("./populate/richTextRelationshipPromise");
const _validate = require("./validate");
const _BlockQuote = require("./field/features/BlockQuote");
const _Blocks = require("./field/features/Blocks");
const _BlocksNode = require("./field/features/Blocks/nodes/BlocksNode");
const _Heading = require("./field/features/Heading");
const _Link = require("./field/features/Link");
const _AutoLinkNode = require("./field/features/Link/nodes/AutoLinkNode");
const _LinkNode = require("./field/features/Link/nodes/LinkNode");
const _Paragraph = require("./field/features/Paragraph");
const _Relationship = require("./field/features/Relationship");
const _RelationshipNode = require("./field/features/Relationship/nodes/RelationshipNode");
const _Upload = require("./field/features/Upload");
const _UploadNode = require("./field/features/Upload/nodes/UploadNode");
const _align = require("./field/features/align");
const _floatingSelectToolbarTextDropdownSection = require("./field/features/common/floatingSelectToolbarTextDropdownSection");
const _html = require("./field/features/converters/html");
const _converter = require("./field/features/converters/html/converter");
const _linebreak = require("./field/features/converters/html/converter/converters/linebreak");
const _paragraph = require("./field/features/converters/html/converter/converters/paragraph");
const _text = require("./field/features/converters/html/converter/converters/text");
const _defaultConverters = require("./field/features/converters/html/converter/defaultConverters");
const _field = require("./field/features/converters/html/field");
const _TestRecorder = require("./field/features/debug/TestRecorder");
const _TreeView = require("./field/features/debug/TreeView");
const _Bold = require("./field/features/format/Bold");
const _InlineCode = require("./field/features/format/InlineCode");
const _Italic = require("./field/features/format/Italic");
const _floatingSelectToolbarSection = require("./field/features/format/common/floatingSelectToolbarSection");
const _strikethrough = require("./field/features/format/strikethrough");
const _subscript = require("./field/features/format/subscript");
const _superscript = require("./field/features/format/superscript");
const _underline = require("./field/features/format/underline");
const _horizontalrule = require("./field/features/horizontalrule");
const _indent = require("./field/features/indent");
const _CheckList = require("./field/features/lists/CheckList");
const _OrderedList = require("./field/features/lists/OrderedList");
const _UnorderedList = require("./field/features/lists/UnorderedList");
const _LexicalPluginToLexical = require("./field/features/migrations/LexicalPluginToLexical");
const _SlateToLexical = require("./field/features/migrations/SlateToLexical");
const _blockquote = require("./field/features/migrations/SlateToLexical/converter/converters/blockquote");
const _heading = require("./field/features/migrations/SlateToLexical/converter/converters/heading");
const _indent1 = require("./field/features/migrations/SlateToLexical/converter/converters/indent");
const _link = require("./field/features/migrations/SlateToLexical/converter/converters/link");
const _listItem = require("./field/features/migrations/SlateToLexical/converter/converters/listItem");
const _orderedList = require("./field/features/migrations/SlateToLexical/converter/converters/orderedList");
const _relationship = require("./field/features/migrations/SlateToLexical/converter/converters/relationship");
const _unknown = require("./field/features/migrations/SlateToLexical/converter/converters/unknown");
const _unorderedList = require("./field/features/migrations/SlateToLexical/converter/converters/unorderedList");
const _upload = require("./field/features/migrations/SlateToLexical/converter/converters/upload");
const _defaultConverters1 = require("./field/features/migrations/SlateToLexical/converter/defaultConverters");
const _index = require("./field/features/migrations/SlateToLexical/converter/index");
const _EditorConfigProvider = require("./field/lexical/config/EditorConfigProvider");
const _loader = require("./field/lexical/config/loader");
const _nodes = require("./field/lexical/nodes");
const _index1 = require("./field/lexical/plugins/SlashMenu/LexicalTypeaheadMenuPlugin/index");
const _types = require("./field/lexical/plugins/SlashMenu/LexicalTypeaheadMenuPlugin/types");
const _canUseDOM = require("./field/lexical/utils/canUseDOM");
const _getDOMRangeRect = require("./field/lexical/utils/getDOMRangeRect");
const _getSelectedNode = require("./field/lexical/utils/getSelectedNode");
const _guard = require("./field/lexical/utils/guard");
const _invariant = require("./field/lexical/utils/invariant");
const _joinClasses = require("./field/lexical/utils/joinClasses");
const _createBlockNode = require("./field/lexical/utils/markdown/createBlockNode");
const _nodeFormat = require("./field/lexical/utils/nodeFormat");
const _point = require("./field/lexical/utils/point");
const _rect = require("./field/lexical/utils/rect");
const _setFloatingElemPosition = require("./field/lexical/utils/setFloatingElemPosition");
const _setFloatingElemPositionForLinkEditor = require("./field/lexical/utils/setFloatingElemPositionForLinkEditor");
const _swipe = require("./field/lexical/utils/swipe");
const _url = require("./field/lexical/utils/url");
const _defaultValue = require("./populate/defaultValue");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
function lexicalEditor(props) {
    let finalSanitizedEditorConfig;
    if (!props || !props.features && !props.lexical) {
        finalSanitizedEditorConfig = (0, _cloneDeep.cloneDeep)(_default.defaultSanitizedEditorConfig);
    } else {
        let features = props.features && typeof props.features === 'function' ? props.features({
            defaultFeatures: (0, _cloneDeep.cloneDeep)(_default.defaultEditorFeatures)
        }) : props.features;
        if (!features) {
            features = (0, _cloneDeep.cloneDeep)(_default.defaultEditorFeatures);
        }
        const lexical = props.lexical;
        finalSanitizedEditorConfig = (0, _sanitize.sanitizeEditorConfig)({
            features,
            lexical: props.lexical ? ()=>Promise.resolve(lexical) : _default.defaultEditorConfig.lexical
        });
    }
    return {
        LazyCellComponent: ()=>// @ts-expect-error
            Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("./cell"))).then((module)=>{
                const RichTextCell = module.RichTextCell;
                return Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("payload/utilities"))).then((module2)=>module2.withMergedProps({
                        Component: RichTextCell,
                        toMergeIntoProps: {
                            editorConfig: finalSanitizedEditorConfig
                        }
                    }));
            }),
        LazyFieldComponent: ()=>// @ts-expect-error
            Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("./field"))).then((module)=>{
                const RichTextField = module.RichTextField;
                return Promise.resolve().then(()=>/*#__PURE__*/ _interop_require_wildcard(require("payload/utilities"))).then((module2)=>module2.withMergedProps({
                        Component: RichTextField,
                        toMergeIntoProps: {
                            editorConfig: finalSanitizedEditorConfig
                        }
                    }));
            }),
        afterReadPromise: ({ field, incomingEditorState, siblingDoc })=>{
            return new Promise((resolve, reject)=>{
                const promises = [];
                if (finalSanitizedEditorConfig?.features?.hooks?.afterReadPromises?.length) {
                    for (const afterReadPromise of finalSanitizedEditorConfig.features.hooks.afterReadPromises){
                        const promise = afterReadPromise({
                            field,
                            incomingEditorState,
                            siblingDoc
                        });
                        if (promise) {
                            promises.push(promise);
                        }
                    }
                }
                Promise.all(promises).then(()=>resolve()).catch((error)=>reject(error));
            });
        },
        editorConfig: finalSanitizedEditorConfig,
        outputSchema: ({ collectionIDFieldTypes, config, field, interfaceNameDefinitions, isRequired, payload })=>{
            let outputSchema = {
                // This schema matches the SerializedEditorState type so far, that it's possible to cast SerializedEditorState to this schema without any errors.
                // In the future, we should
                // 1) allow recursive children
                // 2) Pass in all the different types for every node added to the editorconfig. This can be done with refs in the schema.
                type: (0, _utilities.withNullableJSONSchemaType)('object', isRequired),
                properties: {
                    root: {
                        type: 'object',
                        additionalProperties: false,
                        properties: {
                            type: {
                                type: 'string'
                            },
                            children: {
                                type: 'array',
                                items: {
                                    type: 'object',
                                    additionalProperties: true,
                                    properties: {
                                        type: {
                                            type: 'string'
                                        },
                                        version: {
                                            type: 'integer'
                                        }
                                    },
                                    required: [
                                        'type',
                                        'version'
                                    ]
                                }
                            },
                            direction: {
                                oneOf: [
                                    {
                                        enum: [
                                            'ltr',
                                            'rtl'
                                        ]
                                    },
                                    {
                                        type: 'null'
                                    }
                                ]
                            },
                            format: {
                                type: 'string',
                                enum: [
                                    'left',
                                    'start',
                                    'center',
                                    'right',
                                    'end',
                                    'justify',
                                    ''
                                ]
                            },
                            indent: {
                                type: 'integer'
                            },
                            version: {
                                type: 'integer'
                            }
                        },
                        required: [
                            'children',
                            'direction',
                            'format',
                            'indent',
                            'type',
                            'version'
                        ]
                    }
                },
                required: [
                    'root'
                ]
            };
            for (const modifyOutputSchema of finalSanitizedEditorConfig.features.generatedTypes.modifyOutputSchemas){
                outputSchema = modifyOutputSchema({
                    collectionIDFieldTypes,
                    config,
                    currentSchema: outputSchema,
                    field,
                    interfaceNameDefinitions,
                    isRequired,
                    payload
                });
            }
            return outputSchema;
        },
        populationPromise ({ context, currentDepth, depth, draft, field, findMany, flattenLocales, overrideAccess, populationPromises, req, showHiddenFields, siblingDoc }) {
            // check if there are any features with nodes which have populationPromises for this field
            if (finalSanitizedEditorConfig?.features?.populationPromises?.size) {
                return (0, _richTextRelationshipPromise.richTextRelationshipPromise)({
                    context,
                    currentDepth,
                    depth,
                    draft,
                    editorPopulationPromises: finalSanitizedEditorConfig.features.populationPromises,
                    field,
                    findMany,
                    flattenLocales,
                    overrideAccess,
                    populationPromises,
                    req,
                    showHiddenFields,
                    siblingDoc
                });
            }
            return null;
        },
        validate: (0, _validate.richTextValidateHOC)({
            editorConfig: finalSanitizedEditorConfig
        })
    };
}
