// @payloadcms/richtext-lexical/dist/field/Field.js
'use client';
"use strict";
Object.defineProperty(exports, "__esModule", {
    value: true
});
Object.defineProperty(exports, "default", {
    enumerable: true,
    get: function() {
        return _default;
    }
});
const _forms = require("payload/components/forms");
const _react = /*#__PURE__*/ _interop_require_wildcard(require("react"));
const _reacterrorboundary = require("react-error-boundary");
const _validate = require("../validate");
require("./index.scss");
const _LexicalProvider = require("./lexical/LexicalProvider");
function _getRequireWildcardCache(nodeInterop) {
    if (typeof WeakMap !== "function") return null;
    var cacheBabelInterop = new WeakMap();
    var cacheNodeInterop = new WeakMap();
    return (_getRequireWildcardCache = function(nodeInterop) {
        return nodeInterop ? cacheNodeInterop : cacheBabelInterop;
    })(nodeInterop);
}
function _interop_require_wildcard(obj, nodeInterop) {
    if (!nodeInterop && obj && obj.__esModule) {
        return obj;
    }
    if (obj === null || typeof obj !== "object" && typeof obj !== "function") {
        return {
            default: obj
        };
    }
    var cache = _getRequireWildcardCache(nodeInterop);
    if (cache && cache.has(obj)) {
        return cache.get(obj);
    }
    var newObj = {
        __proto__: null
    };
    var hasPropertyDescriptor = Object.defineProperty && Object.getOwnPropertyDescriptor;
    for(var key in obj){
        if (key !== "default" && Object.prototype.hasOwnProperty.call(obj, key)) {
            var desc = hasPropertyDescriptor ? Object.getOwnPropertyDescriptor(obj, key) : null;
            if (desc && (desc.get || desc.set)) {
                Object.defineProperty(newObj, key, desc);
            } else {
                newObj[key] = obj[key];
            }
        }
    }
    newObj.default = obj;
    if (cache) {
        cache.set(obj, newObj);
    }
    return newObj;
}
const baseClass = 'rich-text-lexical';
const RichText = (props)=>{
    const { name, admin: { className, components: { Error, Label } = {}, condition, description, readOnly, style, width } = {
        className,
        components: {},
        condition,
        description,
        readOnly,
        style,
        width
    }, editorConfig, label, path: pathFromProps, required, validate = (0, _validate.richTextValidateHOC)({
        editorConfig
    }) } = props;
    const path = pathFromProps || name;
    const memoizedValidate = (0, _react.useCallback)((value, validationOptions)=>{
        return validate(value, {
            ...validationOptions,
            props,
            required
        });
    }, // Important: do not add props to the dependencies array.
    // This would cause an infinite loop and endless re-rendering.
    // Removing props from the dependencies array fixed this issue: https://github.com/payloadcms/payload/issues/3709
    [
        validate,
        required
    ]);
    const fieldType = (0, _forms.useField)({
        condition,
        path,
        validate: memoizedValidate
    });
    const { errorMessage, initialValue, setValue, showError, value } = fieldType;
    const classes = [
        baseClass,
        'field-type',
        className,
        showError && 'error',
        readOnly && `${baseClass}--read-only`
    ].filter(Boolean).join(' ');
    const ErrorComp = Error || _forms.Error;
    const LabelComp = Label || _forms.Label;
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: classes,
        key: path,
        style: {
            ...style,
            width
        }
    }, /*#__PURE__*/ _react.default.createElement("div", {
        className: `${baseClass}__wrap`
    }, /*#__PURE__*/ _react.default.createElement(ErrorComp, {
        message: errorMessage,
        showError: showError
    }), /*#__PURE__*/ _react.default.createElement(LabelComp, {
        htmlFor: `field-${path.replace(/\./g, '__')}`,
        label: label,
        required: required
    }), /*#__PURE__*/ _react.default.createElement(_reacterrorboundary.ErrorBoundary, {
        fallbackRender: fallbackRender,
        onReset: ()=>{}
    }, /*#__PURE__*/ _react.default.createElement(_LexicalProvider.LexicalProvider, {
        editorConfig: editorConfig,
        fieldProps: props,
        key: JSON.stringify({
            initialValue,
            path
        }),
        onChange: (editorState)=>{
            let serializedEditorState = editorState.toJSON();
            // Transform state through save hooks
            if (editorConfig?.features?.hooks?.save?.length) {
                editorConfig.features.hooks.save.forEach((hook)=>{
                    serializedEditorState = hook({
                        incomingEditorState: serializedEditorState
                    });
                });
            }
            setValue(serializedEditorState);
        },
        path: path,
        readOnly: readOnly,
        value: value
    })), /*#__PURE__*/ _react.default.createElement(_forms.FieldDescription, {
        description: description,
        path: path,
        value: value
    })));
};
function fallbackRender({ error }) {
    // Call resetErrorBoundary() to reset the error boundary and retry the render.
    return /*#__PURE__*/ _react.default.createElement("div", {
        className: "errorBoundary",
        role: "alert"
    }, /*#__PURE__*/ _react.default.createElement("p", null, "Something went wrong:"), /*#__PURE__*/ _react.default.createElement("pre", {
        style: {
            color: 'red'
        }
    }, error.message));
}
const _default = (0, _forms.withCondition)(RichText);
